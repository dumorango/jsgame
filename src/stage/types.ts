import { Sprite } from "./../sprite";

export type Position = {
  x: number;
  y: number;
};

export type Stage<Tile> = {
  layers: TiledLayer<Tile>[];
};

export type TiledLayer<Tile> = {
  imageFile: string;
  tileMap: Map<Tile, Position>;
  layout: TiledLayout<Tile>;
};

export type LoadedStage<Tile> = {
  layers: LoadedTiledLayer<Tile>[];
};

export type LoadedTiledLayer<Tile> = {
  img: HTMLImageElement;
  tileMap: Map<Tile, Position>;
  layout: TiledLayout<Tile>;
};

export type TiledLayout<Tile> = {
  patterns: TiledPatternPlacement<Tile>[];
  tileSpecs: TileSpecs;
};

export type TiledPatternPlacement<Tile> = {
  position: Position;
  pattern: TiledPattern<Tile>;
};

export type TiledPattern<Tile> = Tile[][];

export type TilePlacement<Tile> = {
  tile: Tile;
  position: Position;
};

export type TileSpecs = {
  size: number;
  padding: number;
};

export type SpritePlacement = {
  sprite: Sprite;
  position: Position;
};
