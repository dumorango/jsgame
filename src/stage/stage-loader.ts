
import {
  EnemySpawnPoint,
  StageState,
  StageDesign,
  PlayerState,
  Position,
  EnemyState,
  TiledLayerDesign,
} from "./types";
import { getPatternState } from './patterns';
import { mario } from "../sprite";
import { makeGoomba } from "./enemies";

const getPlayerInitialState = (
  playerSpawnPointPosition: Position
): PlayerState => {
  return {
    movement: {
      position: playerSpawnPointPosition,
      velocity: { x: 0, y: 0 },
      acceleration: { x: 0, y: 0 },
    },
    sprite: mario.stalled,
    isDead: false,
  };
};

const getEnemyState = (enemySpawnPoint: EnemySpawnPoint): EnemyState => {
  if (enemySpawnPoint.type === "GOOMBA") {
    return makeGoomba(enemySpawnPoint.position);
  } else {
    throw new Error(`Enemy type not found: ${enemySpawnPoint.type}`);
  }
};

function* getPatternsState(layers: TiledLayerDesign[]) {
  for(const layer of layers) {
    for(const patternSpawn of layer.patterns) {
      yield getPatternState(patternSpawn)
    }
  }
}

export const getStageState = (stageDesign: StageDesign): StageState => {
  return {    
    mario: getPlayerInitialState(stageDesign.player.position),
    enemies: stageDesign.enemies.map(getEnemyState),
    patterns: [...getPatternsState(stageDesign.layers)]
  };
};
