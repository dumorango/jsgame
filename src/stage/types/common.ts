import { Sprite } from "../../sprite";

export type Position = {
  x: number;
  y: number;
};

export type Size = {
  height: number;
  width: number;
};

export type SpritesImage = string;

export type SpritePlacement = {
  sprite: Sprite;
  position: Position;
};

