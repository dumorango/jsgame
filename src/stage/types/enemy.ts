
export enum EnemyType {
  GOOMBA = "GOOMBA"
}

export type EnemyPlacement = {
  enemy: EnemyType;
  position: Position;
};
