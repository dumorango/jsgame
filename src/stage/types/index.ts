export * from "./design";
export * from "./enemy";
export * from "./common";
export * from "./state";
export * from "./tile";
export * from "./pattern";
export * from "./trait";