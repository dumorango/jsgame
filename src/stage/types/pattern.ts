import { Size } from ".";

export type SkyPattern = {
  type: PatternType.SKY;
  size: Size;
};

export type GroundPattern = {
  type: PatternType.GROUND;
  length: number;
};

export type PipePattern = {
  type: PatternType.PIPE;
  height: number;
};

export type BrickPattern = {
  type: PatternType.BRICK;
};

export enum PatternType {
  GROUND = "GROUND",
  SKY = "SKY",
  PIPE = "PIPE",
  BRICK = "BRICK",
}

export type Pattern = SkyPattern | GroundPattern | PipePattern | BrickPattern;
