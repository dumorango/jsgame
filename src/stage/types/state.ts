import { EnemyType, SpritesImage, Pattern } from ".";
import { Sprite } from "./../../sprite/types";
import { Movement } from "../../movement";
import { Trait } from "./trait";

export type StageState = {  
  mario: PlayerState;
  enemies: EnemyState[];
  patterns: PatternState[];
};

export type EnemyState = {
  type: EnemyType;
  movement: Movement;
  traits: Trait[];
  isDead: boolean;
  sprite: Sprite;
};

export type Enemies = {
  enemies: EnemyState[];
  image: SpritesImage;
};

export type PlayerState = {
  movement: Movement;
  isDead: boolean;
  sprite: Sprite;
};

export type PatternState = {
  movement: Movement;
  pattern: Pattern;  
  traits?: Trait[]  
};
