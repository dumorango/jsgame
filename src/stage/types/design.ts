import { Pattern } from './../types';
import { EnemyType } from "./enemy";
import { Position } from "./common";

export type StageDesign = {
  layers: TiledLayerDesign[];
  enemies: EnemySpawnPoint[];
  player: PlayerSpawnPoint;
};

export type TiledLayerDesign = {
  patterns: PatternSpawnPoint[];
};

export type PatternSpawnPoint = {
  pattern: Pattern,
  position: Position,
}

export type EnemySpawnPoint = {
  type: EnemyType;
  position: Position;
};

export type PlayerSpawnPoint = {  
  position: Position;
};
