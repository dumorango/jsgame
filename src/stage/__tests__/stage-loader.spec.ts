import { Stage } from "./../types";
import { loadStageLayersImage } from "../stage-loader";
import { loadImage } from "../../loaders";

jest.mock("../../loaders");

describe("The stage loader module", () => {
  it("should load the stage layers images", async () => {
    const imageFile = "imagepath.png";
    const img = document.createElement("img");
    const layers = [
      {
        imageFile,
      },
    ];
    const stage = {
      layers,
    } as Stage<any>;
    (loadImage as jest.Mock).mockResolvedValue(img);
    const response = await loadStageLayersImage(stage);
    expect(response).toMatchObject({
      layers: [
        {
          img,
        },
      ],
    });
  });
});
