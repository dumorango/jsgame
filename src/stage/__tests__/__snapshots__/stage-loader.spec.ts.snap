// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The stage loader module should load the stage state from a stage design 1`] = `
Object {
  "enemies": Array [
    Object {
      "isDead": false,
      "movement": Object {
        "acceleration": Object {
          "x": -0.15,
          "y": 0,
        },
        "position": Object {
          "x": 0,
          "y": 0,
        },
        "velocity": Object {
          "x": 0,
          "y": 0,
        },
      },
      "sprite": Object {
        "coordinates": Object {
          "position": Object {
            "x": 2,
            "y": 155,
          },
          "size": Object {
            "height": 16,
            "width": 16,
          },
        },
        "image": "ENEMIES",
      },
      "traits": Array [
        "SIDEWAYS_WALKER",
        "BOUCEABLE",
      ],
      "type": "GOOMBA",
    },
  ],
  "mario": Object {
    "isDead": false,
    "movement": Object {
      "acceleration": Object {
        "x": 0,
        "y": 0,
      },
      "position": Object {
        "x": 0,
        "y": 0,
      },
      "velocity": Object {
        "x": 0,
        "y": 0,
      },
    },
    "sprite": Object {
      "coordinates": Object {
        "position": Object {
          "x": 216,
          "y": 89,
        },
        "size": Object {
          "height": 15,
          "width": 14,
        },
      },
      "image": "MARIO",
    },
  },
  "patterns": Array [
    Object {
      "movement": Object {
        "acceleration": Object {
          "x": 0,
          "y": 0,
        },
        "position": Object {
          "x": 0,
          "y": 0,
        },
        "velocity": Object {
          "x": 0,
          "y": 0,
        },
      },
      "pattern": Object {
        "type": "BRICK",
      },
      "traits": Array [
        "SOLID",
        "BREAKABLE",
      ],
    },
  ],
}
`;
